src/main/java/com/avispl/symphony/dal/avdevices/encoderdecoder/haivision/x4decoder/HaivisionX4DecoderCommunicator.java

```diff
@@ -0,0 +1,755 @@
+					}
+					// Update invalid port number
+					if (setPortNumber != null) {
+						updateInvalidAdapterProperties(stats, setPortNumber, FilteringErrorGroup.STREAM_STATUS_AND_PORT_NUMBER.getName(), StreamConfigMetric.PORT.getName());
+					}
+				} else {
+					updateStreamStatisticsFailedMonitor(failedMonitor);
+				}
+			} else {
+				updateStreamStatisticsFailedMonitor(failedMonitor);
+			}
+		} catch (Exception e) {
+			updateStreamStatisticsFailedMonitor(failedMonitor);
+		}
+	}
+
+	/**
+	 * Counting metric group is failed to monitor
+	 *
+	 * @return number failed monitoring metric group in the metric
+	 */
+	private int getNoOfFailedMonitorMetricGroup() {
+		int noOfFailedMonitorMetric = 3;
+		for (MonitoringMetricGroup metric : MonitoringMetricGroup.values()) {
+			if (metric.isFailedMonitorCheck()) {
+				noOfFailedMonitorMetric++;
+			}
+		}
+		return noOfFailedMonitorMetric;
+	}
+
+
+	/**
+	 * This method is used to populate all monitoring properties:
+	 * <li>Decoders statistic</li>
+	 * <li>Streams statistic</li>
+	 *
+	 * @param stats list statistic property
+	 * @throws ResourceNotReachableException when failedMonitor said all device monitoring data are failed to get
+	 */
+	private void populateDecoderMonitoringMetrics(Map<String, String> stats) {
+		Objects.requireNonNull(stats);
+
+		if (!StringUtils.isNullOrEmpty(getPassword()) && !StringUtils.isNullOrEmpty(getLogin())) {
+			retrieveSessionFromDecoder();
+			retrieveUserRoleBasedFromDecoder();
+		} else {
+			this.authenticationCookie.setSessionID(DecoderConstant.AUTHORIZED);
+		}
+		// Retrieving all device info
+		retrieveDeviceInfo(stats);
+
+		// Retrieving all decoders stats
+		for (int decoderID = 0; decoderID < 4; decoderID++) {
+			retrieveDecoderStats(stats, decoderID);
+		}
+
+		// Retrieving all streams stats
+		retrieveStreamStats(stats);
+
+		if (failedMonitor.size() == getNoOfFailedMonitorMetricGroup()) {
+			authenticationCookie.setSessionID(null);
+			StringBuilder errBuilder = new StringBuilder();
+			for (Map.Entry<String, String> failedMetric : failedMonitor.entrySet()) {
+				errBuilder.append(failedMetric.getValue());
+				errBuilder.append(DecoderConstant.NEXT_LINE);
+			}
+			throw new ResourceNotReachableException(errBuilder.toString());
+		}
+	}
+
+	/**
+	 * This method is used to handle  input from adapter properties and convert it to Set of String  for control
+	 *
+	 * @return Set<String> is the Set of String of filter element
+	 */
+	public Set<String> handleAdapterPropertiesInputFromUser(String input) {
+		String[] listAdapterPropertyElement = input.split(String.valueOf(DecoderConstant.COMMA));
+		StringBuilder errorMessages = new StringBuilder();
+
+		// Remove start and end spaces of each gain
+		Set<String> setAdapterPropertiesElement = new HashSet<>();
+		for (int i = 0; i < listAdapterPropertyElement.length; ++i) {
```

Comment:
there are nothing specific with index handling. You should use for each instead of this one.

Reply:
I fixed it